osquery/tables/system/windows/battery.cpp

```diff
@@ -0,0 +1,261 @@
+      // battery in order to convert from the mWh units that
+      // Microsoft provides to match the mAh units that the battery
+      // table already uses for macOS.
+      const int designedVoltage = 12;
+      row["max_capacity"] = INTEGER(bi.FullChargedCapacity / designedVoltage);
+      row["designed_capacity"] = INTEGER(bi.DesignedCapacity / designedVoltage);
+      if (bi.CycleCount != 0) {
+        row["cycle_count"] = INTEGER(bi.CycleCount);
+      }
+
+      // Query the battery power status.
+      BATTERY_WAIT_STATUS bws = {0};
+      bws.BatteryTag = bqi.BatteryTag;
+      BATTERY_STATUS bs;
+      if (DeviceIoControl(hBattery,
+                          IOCTL_BATTERY_QUERY_STATUS,
+                          &bws,
+                          sizeof(bws),
+                          &bs,
+                          sizeof(bs),
+                          &dwOut,
+                          nullptr)) {
+        // https://learn.microsoft.com/en-us/windows/win32/power/battery-wait-status-str
+        if (bs.PowerState & BATTERY_POWER_ON_LINE) {
+          row["state"] = "AC Power";
+          row["charging"] = INTEGER((bs.PowerState & BATTERY_CHARGING) > 0);
+        } else if (bs.PowerState & BATTERY_DISCHARGING) {
+          row["state"] = "Battery Power";
+          row["charging"] = INTEGER(0);
+        }
+        row["charged"] = INTEGER(bs.Capacity == bi.FullChargedCapacity);
+        row["current_capacity"] = INTEGER(bs.Capacity / designedVoltage);
+        row["voltage"] = INTEGER(bs.Voltage);
+        if (bs.Voltage > 0) {
+          row["amperage"] = INTEGER((1000 * static_cast<int>(bs.Rate)) /
+                                    static_cast<int>(bs.Voltage));
+        } else {
+          LOG(WARNING) << "Battery table read a voltage of 0.";
+        }
+        if (bs.Capacity != bi.FullChargedCapacity && bs.Rate > 0) {
+          row["minutes_to_full_charge"] =
+              INTEGER(60 * (bi.FullChargedCapacity - bs.Capacity) / bs.Rate);
+        }
+      }
+
+      SYSTEM_POWER_STATUS sps;
+      if (GetSystemPowerStatus(&sps)) {
+        if (sps.BatteryLifePercent != -1) {
+          row["percent_remaining"] =
+              INTEGER((unsigned int)sps.BatteryLifePercent);
+        }
+        if (sps.BatteryLifeTime != -1) {
+          row["minutes_until_empty"] =
+              INTEGER(sps.BatteryLifeTime / 60); // convert seconds to minutes
+        }
+      } else {
+        LOG(WARNING) << "Failed to get system power status";
+      }
+
+      row["manufacturer"] = batteryQueryInformationString(
+          hBattery, bqi.BatteryTag, BatteryManufactureName);
+
+      row["serial_number"] = batteryQueryInformationString(
+          hBattery, bqi.BatteryTag, BatterySerialNumber);
+
+      row["model"] = batteryQueryInformationString(
+          hBattery, bqi.BatteryTag, BatteryDeviceName);
+
+      // Once we find one battery, no need to do anything
+      // else

```

Comment:
Shouldn't we return all the batteries? There are some laptops with 2 batteries. Also, for some laptops you can get an external battery, like https://www.amazon.com/593553-001-593554-001-Extended-Pavilion-CQ42/dp/B07SGHM9RR

Reply:
Yeah, good point. I've never encountered such a setup (and don't have one to test on) but it's pretty easy to allow this code to support it.

Reply:
Updated to support this but I'm not able to test it. Code still works for 0/1 batteries.